File: Leetcode/C++/687. Longest Univalue Path/solution.cpp
// https://leetcode.com/problems/longest-univalue-path/description/

#include <algorithm>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    int length = 0;
    int longestUnivaluePath(TreeNode *root)
    {
        search(root, 0);
        return length;
    }

    int search(TreeNode *root, int parentVal)
    {
        if (root == NULL)
        {
            return 0;
        }

        int left = search(root->left, root->val), right = search(root->right, root->val);
        length = max(length, left + right);
        // If parent node has same value add current node to path else reset path
        return parentVal == root->val ? max(left, right) + 1 : 0;
    }
};

File: Leetcode/Python/687. Longest Univalue Path/solution.py
# https://leetcode.com/problems/longest-univalue-path/


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    answer = 0

    def longestUnivaluePath(self, root: Optional[TreeNode]) -> int:
        def dfs(node):
            if not node:
                return 0

            left, right = dfs(node.left), dfs(node.right)

            if node.left and node.left.val == node.val:
                left += 1
            else:
                left = 0

            if node.right and node.right.val == node.val:
                right += 1
            else:
                right = 0

            self.answer = max(self.answer, left + right)

            return max(left, right)

        dfs(root)
        return self.answer
